File: credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

long long card;
int main()
{
    do
    {
        card = get_long_long("Number:");
    }
    while (card == 0);

    // first we check the length
    long long length = (long long) log10((double) card) + 1;
    // printf("%lld\n", card);
    // printf("%lld\n", length);
    if (length != 13 && length != 15 && length != 16)
    {
        printf("INVALID\n");
        return 0;
    }

    // second we check which card it could be
    string OUTPUT = "INVALID";
    // printf("%lld\n", card / 1000000000000000);
    if ((length == 13) && (card / 1000000000000 == 4))
    {
        OUTPUT = "VISA";
    }
    else if ((length == 15) && ((card / 10000000000000 == 34) || (card / 10000000000000 == 37)))
    {
        OUTPUT = "AMEX";
    }
    else if (length == 16)
    {
        if (card / 1000000000000000 == 4)
        {
            // printf("visa baby\n");
            OUTPUT = "VISA";
        }
        else if (card / 100000000000000 >= 51 && card / 100000000000000 <= 55)
        {
            OUTPUT = "MASTERCARD";
        }
    }
    else
    {
        OUTPUT = "INVALID";
    }

    // third we calculate luhn's sum
    long long luhn_sum = 0;
    for (int i = 1; i <= length; i++)
    {
        if (i % 2 == 1)
        {
            luhn_sum = luhn_sum + (card % 10);
            card = card / 10;
        }
        else if (i % 2 == 0)
        {
            long long buffer = 2 * (card % 10);
            luhn_sum = luhn_sum + (buffer % 10) + (buffer / 10);
            card = card / 10;
        }
    }
    // printf("%lld\n", luhn_sum);

    // fourth we output the type
    if (luhn_sum % 10 == 0)
    {
        printf("%s\n", OUTPUT);
        return 0;
    }
    else
    {
        printf("INVALID\n");
        return 0;
    }
}

File: sentimental-credit/credit.py
from cs50 import get_int
import math


# this is the most bruteforce solution i have ever done
# it could be optimized much much further but i still dont know the correct syntax for that
def main():
    card = -1
    # get card number from user
    while card <= 0:
        card = get_int("input card number: ")
    
    card = str(card)
    length = length_check(card)
    card_type = type_of_card(card, length)
    checksum(card, length, card_type)
    
    
def checksum(card, length, card_type):
    # i need to count from behind, if its an odd digit i should simply add it to SUM, if its even i should multiply it by
    # two then add the digits of that multiplication, before adding that to the SUM
    
    # i dont know how to loop through a string from behind so instead if the string is odd lengthed, first digit is treated
    # as an odd digit, if string length is even, first digit is treated as even
    counter = 0
    if length % 2 != 0:
        counter += 1
    SUM = 0
    for i in card:
        if counter % 2 == 0:
            SUM = SUM + (2 * int(i)) % 10 + int((2 * int(i)) / 10)
            counter += 1
        else:
            SUM = SUM + int(i)
            counter += 1
    
    if (SUM % 10 == 0):
        print(card_type)
    else:
        print("INVALID")
        exit()


def type_of_card(card, length):
    # this function returns the expected card type based on length of card number, and the starting digits
    # if the card cannot be a real card, it exits the program without returning anything
    if length == 15 and (int(card[0:2:1]) == 34 or int(card[0:2:1]) == 37):
        return "AMEX"
        
    elif length == 16 and (int(card[0:2:1]) in [51, 52, 53, 54, 55]):
        return "MASTERCARD"
    
    elif (length == 13 or length == 16) and (int(card[0:1:1]) == 4):
        return "VISA"
    else:
        print("INVALID")
        exit()


def length_check(card):
    # this function returns length of card number, it is useful for simply exiting program early when card number length is invalid
    if len(card) in [13, 15, 16]:
        return len(card)
    else:
        print("INVALID")
        exit()


main()